utils/src/main/java/com/cloud/utils/net/cidr/CIDR4.java

```diff
@@ -0,0 +1,172 @@
+//
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+//
+
+package com.cloud.utils.net.cidr;
+
+import java.net.Inet4Address;
+import java.net.Inet6Address;
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+
+import org.apache.commons.lang.NotImplementedException;
+import org.apache.log4j.Logger;
+
+import com.cloud.utils.net.NetUtils;
+
+/**
+ */
+public class CIDR4 implements CIDR {
+    protected final static Logger s_logger = Logger.getLogger(CIDR4.class);
+
+    protected InetAddress baseAddress;
+    protected int cidrMask;
+    private int addressInt;
+    private final int addressEndInt;
+
+    protected CIDR4(Inet4Address newaddr, int mask) {
+        cidrMask = mask;
+        addressInt = ipv4AddressToInt(newaddr);
+        int newmask = ipv4PrefixLengthToMask(mask);
+        addressInt &= newmask;
+        try {
+            baseAddress = intToIPv4Address(addressInt);
+        } catch (UnknownHostException e) {
+            // this should never happen
+        }
+        addressEndInt = addressInt + ipv4PrefixLengthToLength(cidrMask) - 1;
+    }
+
+    public int compareTo(CIDR arg) {
+        if (arg instanceof CIDR6) {
+            throw new NotImplementedException("Not implemented for CIDR6");
+        }
+        CIDR4 o = (CIDR4)arg;
+        if (o.addressInt == addressInt && o.cidrMask == cidrMask) {
+            return 0;
+        }
+        if (o.addressInt < addressInt) {
+            return 1;
+        }
+        if (o.addressInt > addressInt) {
+            return -1;
+        }
+        if (o.cidrMask < cidrMask) {
+            // greater Mask means less IpAddresses so -1
+            return -1;
+        }
+        return 1;
+    }
+
+    @Override
+    public boolean isNull() {
+        return false;
+    }
+
+    @Override
+    public boolean contains(InetAddress inetAddress) {
+        if (inetAddress == null) {
+            throw new NullPointerException("inetAddress");
```

Comment:
``NullPointerException`` should only be thrown by VM when a null point is dereferenced.  Convert to a ``Preconditions.checkArgument`` check to simplify the code and ensure the correct exception type is thrown.